File: packages/dart/lib/src/affix/attach.dart
import 'package:randino/src/affix/data/index.dart';
import 'package:randino/src/internal/utils.dart';

/// One value with a **freshly drawn** token attached, on the side [join] puts
/// it. Internal — the four public functions differ by that one line and by
/// whether they map over a list.
String attachOne(
  String value,
  int length,
  String separator,
  String? charset,
  String Function(String value, String token, String separator) join,
) => join(
  value,
  randToken(
    clampInt(length, 1, affixLengthMax),
    (charset == null || charset.isEmpty) ? affixCharset : charset,
  ),
  separator,
);

/// The token goes behind the value.
String appendToken(String value, String token, String separator) => '$value$separator$token';

/// The token goes in front of it.
String prependToken(String value, String token, String separator) => '$token$separator$value';

File: packages/javascript/lib/affix/attach.ts
// Shared by `randSuffix` and `randPrefix`. Internal — the two of them differ by
// one line, which is the side the token lands on.

import { clamp, randToken } from '../_internal/utils.js';
import type { RandAffixOptions } from '../_types/global.js';
import {
	AFFIX_CHARSET,
	AFFIX_LENGTH_DEFAULT,
	AFFIX_LENGTH_MAX,
	AFFIX_SEPARATOR_DEFAULT
} from './data/index.js';

/**
 * Attaches a **freshly drawn** token to one string, or to every string in an
 * array — one token each rather than one for the batch, which is the whole
 * point of passing a generator's output here.
 */
export function attach(
	value: string | string[],
	options: RandAffixOptions,
	join: (value: string, token: string, separator: string) => string
): string | string[] {
	const length = clamp(Math.floor(options.length ?? AFFIX_LENGTH_DEFAULT), 1, AFFIX_LENGTH_MAX);
	const charset = options.charset || AFFIX_CHARSET;
	const separator = options.separator ?? AFFIX_SEPARATOR_DEFAULT;
	const one = (item: string) => join(item, randToken(length, charset), separator);

	return Array.isArray(value) ? value.map(one) : one(value);
}
